autoload/lsp/omni.vim

```diff
@@ -332,141 +297,58 @@ function! lsp#omni#get_vim_completion_item(...) abort
+    let l:user_data[s:user_data_completion_item_key] = a:item
 
-    if has_key(l:user_data, s:user_data_additional_edits_key)
-        let l:all_text_edits += l:user_data[s:user_data_additional_edits_key]
-    endif
+    return l:user_data
+endfunction
 
-    " save cursor position in a mark, vim will move it appropriately when
-    " applying edits
-    let l:saved_mark = getpos("'a")
-    " move to end of newText but in two steps (as column may not exist yet)
-    let [l:pos, l:col_offset] = s:get_cursor_pos_and_edit_length(l:text_edit)
-    call setpos("'a", l:pos)
-
-    " apply textEdits
-    if !empty(l:all_text_edits)
-        call lsp#utils#text_edit#apply_text_edits(lsp#utils#get_buffer_uri(), l:all_text_edits)
-        " When user typed something character while popup menu is shwon, vim
-        " insert typed-character after CompleteDone occured. but the character
-        " should not be duplicated since the textEdit include the character.
-        " this remove the following character.
-        if l:snippet_marker_pos != -1
-            let l:oldpos = line('.')
-            let l:oldline = getline('.')
-            call timer_start(1, {_-> [
-        \    setline(l:oldpos, l:oldline),
-        \    execute('redraw', 1),
-        \    execute('doautocmd User lsp_complete_done', 1),
-        \] })
-        endif
-        return
+function! lsp#omni#extract_user_data_from_completed_item(completed_item) abort
+    " the item has no user_data.
+    if !has_key(a:completed_item, 'user_data')
+        return {}
     endif
 
-    let l:pos = getpos("'a")
-    if l:snippet_marker_pos >= 0
-        let l:pos[2] += l:snippet_marker_pos
-    else
-        let l:pos[2] += l:col_offset
+    " try to decode user_data.
+    try
+        let l:user_data = json_decode(a:completed_item.user_data)
+    catch /.*/
+        let l:user_data = {}
+    endtry
+    if empty(l:user_data)
+        return {}
     endif
-    call setpos("'a", l:saved_mark)
-    call setpos('.', l:pos)
-
-    doautocmd User lsp_complete_done
-endfunction
-
-function! s:expand_range(text_edit, expand_length) abort
-    let l:expanded_text_edit = a:text_edit
-    let l:expanded_text_edit['range']['end']['character'] += a:expand_length
 
-    return l:expanded_text_edit
-endfunction
-
-function! s:get_cursor_pos_and_edit_length(text_edit) abort
-    if !empty(a:text_edit)
-        let l:start = a:text_edit['range']['start']
-        let l:line = l:start['line'] + 1
-        let l:char = l:start['character']
-        let l:col = lsp#utils#to_col('%', l:line, l:char)
-        let l:length = len(a:text_edit['newText'])
-        let l:pos = [0, l:line, l:col, 0]
-    else
-        let l:length = 0
-        let l:pos = getpos('.')
+    " user_data has no managed keys.
+    if !has_key(l:user_data, s:user_data_server_name_key)
+                \ || !has_key(l:user_data, s:user_data_completion_item_key)
+        return {}
     endif
 
-    return [l:pos, l:length]
+    return {
```

Comment:
Why not just return `l:user_data`? That it is consistent and we don't need to remember that sometimes it is `server_name` and sometimes it is `vim-lsp/serverName`.

Reply:
Thanks for your advice.
I've fixed it.